backend/infrahub/graphql/auth/query_permission_checker/object_permission_checker.py

```diff
@@ -119,34 +119,53 @@ class PermissionManagerPermissionChecker(GraphQLQueryPermissionCheckerInterface)
     permission_required = GlobalPermission(
         action=GlobalPermissions.MANAGE_PERMISSIONS.value, decision=PermissionDecision.ALLOW_ALL.value
     )
+    # Map kinds and the relationship to protect from being read
+    kind_relationship_to_check = {
+        InfrahubKind.ACCOUNTROLE: "permissions",
+        InfrahubKind.BASEPERMISSION: "roles",
+        InfrahubKind.GLOBALPERMISSION: "roles",
+        InfrahubKind.OBJECTPERMISSION: "roles",
+    }
 
     async def supports(self, db: InfrahubDatabase, account_session: AccountSession, branch: Branch) -> bool:  # noqa: ARG002
         return config.SETTINGS.main.allow_anonymous_access or account_session.authenticated
 
     async def check(
         self,
-        db: InfrahubDatabase,
+        db: InfrahubDatabase,  # noqa: ARG002
         account_session: AccountSession,  # noqa: ARG002
         analyzed_query: InfrahubGraphQLQueryAnalyzer,
         query_parameters: GraphqlParams,
-        branch: Branch,
+        branch: Branch,  # noqa: ARG002
     ) -> CheckerResolution:
-        is_permission_operation = False
-        kinds = analyzed_query.query_report.impacted_models
-
-        for kind in kinds:
-            schema = get_schema(db=db, branch=branch, node_schema=kind)
-            if is_permission_operation := kind in (
-                InfrahubKind.BASEPERMISSION,
-                InfrahubKind.GLOBALPERMISSION,
-                InfrahubKind.OBJECTPERMISSION,
-            ) or (isinstance(schema, NodeSchema) and InfrahubKind.BASEPERMISSION in schema.inherit_from):
-                break
-
-        if not is_permission_operation:
-            return CheckerResolution.NEXT_CHECKER
-
-        query_parameters.context.active_permissions.raise_for_permission(permission=self.permission_required)
+        for kind, relationship in self.kind_relationship_to_check.items():
+            if (
+                kind in analyzed_query.query_report.requested_read
+                and relationship in analyzed_query.query_report.requested_read[kind].relationships
+            ):
+                query_parameters.context.active_permissions.raise_for_permission(permission=self.permission_required)
+
+        for query in analyzed_query.query_report.queries:
+            if not query.infrahub_model:
+                continue
+
+            if (
+                isinstance(query.infrahub_model, NodeSchema)
+                and InfrahubKind.BASEPERMISSION in query.infrahub_model.inherit_from
+                and query.operation == GraphQLOperation.MUTATION
+            ):
+                query_parameters.context.active_permissions.raise_for_permission(permission=self.permission_required)
+
+            if query.infrahub_model.kind == InfrahubKind.ACCOUNTROLE and query.operation == GraphQLOperation.MUTATION:
```

Comment:
I would say that for mutations we can probably just forbid anything that touches accountroles or permissions in general and not look at the individual fields?